add lmdb version 0.9.23

chincheta0815 opened the pull request:
needed to build future builds with samba.
as of now, lmdb support has to be explicitly disabled for building samba.
having this will hopefully avoid that.

Please look a little bit deeper in this as it is my first completely written pkg.
The build routine for lmdb is also quite "special".

Thanks in advance.

chincheta0815 committed:
add lmdb version 0.9.23

Mno-hime commented on components/database/lmdb/Makefile:
```diff
@@ -0,0 +1,46 @@
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
```
Your copyright statement here. Same for lmdb.p5m.

chincheta0815 replied:
Should be okay now.

Mno-hime commented on components/database/lmdb/Makefile:
```diff
@@ -0,0 +1,46 @@
+COMPONENT_FMRI= database/lmdb
+COMPONENT_CLASSIFICATION=	System/Databases
+COMPONENT_LICENSE=	OpenLDAP Public License
+COMPONENT_LICENSE_FILE=	$(COMPONENT_NAME).license
```
More `tab` alignment in this section, please.

chincheta0815 replied:
Should be okay now.

Mno-hime commented on components/database/lmdb/Makefile:
```diff
@@ -0,0 +1,46 @@
+
+COMPONENT_INSTALL_ARGS += bindir=$(USRBINDIR.$(BITS)) \
+	includedir=$(USRINCDIR) libdir=$(USRLIBDIR.$(BITS)) \
+	mandir=$(USRSHAREMANDIR)
```
`libdir` on new line, please.

chincheta0815 replied:
Should be okay now.

Mno-hime commented on components/database/lmdb/Makefile:
```diff
@@ -0,0 +1,46 @@
+
+build: $(BUILD_32_and_64)
+
+test: $(NO_TEST)
```
There's a `test` target in the upstream Makefile. Can we leverage it?

Swap test and install targets, please.

chincheta0815 replied:
Swap done.
Concerning the test: the output of the test target seems to differ on purpose. Up to now I did not get any info on how to decide whether the test passed okay or not. All I got is that the program runs. In the end one simply needs the libs. Maybe this helps until I get some info from the devs.

Mno-hime replied:
Fedora (https://src.fedoraproject.org/rpms/lmdb/tree/master) shows, how to run them, doing just that and thus seeing that tests did not explode, should be fine for now.

Mno-hime commented on components/database/lmdb/manifests/sample-manifest.p5m:
```diff
@@ -0,0 +1,9 @@
+file path=usr/share/man/man1/mdb_copy.1
+file path=usr/share/man/man1/mdb_dump.1
+file path=usr/share/man/man1/mdb_load.1
+file path=usr/share/man/man1/mdb_stat.1
```
This file should be created by `gmake sample-manifest`. Then compare it with lmdb.p5m.

chincheta0815 replied:
I created it using "gmake sample-manifest MANIFEST_TEMPLATE=" as written in the wiki. Was that wrong? Is there some kind of better description?

Mno-hime replied:
That sound to me like a outdated information. Pure `gmake sample-manifest` makes it as expected with the default template.

Mno-hime commented on components/database/lmdb/lmdb.p5m:
```diff
@@ -0,0 +1,33 @@
+file path=usr/include/lmdb.h
+file path=usr/lib/$(MACH64)/liblmdb.a
+file path=usr/lib/$(MACH64)/liblmdb.so
+file path=usr/lib/liblmdb.a
```
Do we need to ship .a files?

chincheta0815 replied:
Hopefully we won't need them.
I will patch the Makefile accordingly and push an update.

Mno-hime replied:
That will work, though note that just not having them here in the p5m file ensures it's not in the resulting package.

Mno-hime commented on components/database/lmdb/lmdb.p5m:
```diff
@@ -0,0 +1,33 @@
+
+file path=usr/include/lmdb.h
+file path=usr/lib/$(MACH64)/liblmdb.a
+file path=usr/lib/$(MACH64)/liblmdb.so
```
Strangely, .so libraries are not versioned. Can you add a bogus one, please? openSUSE (https://build.opensuse.org/package/show/systemsmanagement/lmdb) and Fedora (https://src.fedoraproject.org/rpms/lmdb/tree/master) seems to have opinion on that.

pyhalov commented on components/database/lmdb/Makefile:
```diff
@@ -0,0 +1,50 @@
+COMPONENT_BUILD_ARGS += LDFLAGS="$(LDFLAGS)"
+
+COMPONENT_INSTALL_ARGS += bindir=$(USRBINDIR.$(BITS)) \
+	includedir=$(USRINCDIR) \
```
Can we compose COMPONENT_INSTALL_ARGS like COMPONENT_BUILD_ARGS?
```
COMPONENT_INSTALL_ARGS += bindir=$(USRBINDIR.$(BITS))
COMPONENT_INSTALL_ARGS += includedir=$(USRINCDIR)
...
````

Mno-hime commented on components/database/lmdb/Makefile:
```diff
@@ -0,0 +1,55 @@
+
+test: $(TEST_32_and_64)
+
+install: $(INSTALL_32_and_64)
```
Swap `install` and `test` targets, please.

chincheta0815 replied:
Sorry, no idea why I always do that wrong

chincheta0815 commented:
Okay, here is a new version with hopefully all changes included.
Concerning the test: I hope they are okay that way.
The devs told me that the test output is nothing useful for a simple user. I hope a simple "well it runs"-test is okay.
The verison of the lib should also be okay now. I added a symlink.

Mno-hime commented:
@chincheta0815 Once you finish those two remaining nits, squash commits, looks a bit hairy right now.

chincheta0815 commented:
> 
> 
> @chincheta0815 Once you finish those two remaining nits, squash commits, looks a bit hairy right now.

Which nits do you mean? I thought I am finished ;o)

Mno-hime commented:
@chincheta0815 I mean this: https://github.com/OpenIndiana/oi-userland/pull/4726#pullrequestreview-195881928. It's suggested to be like this:

```Makefile
COMPONENT_BUILD_ARGS += CC="$(CC)"
COMPONENT_BUILD_ARGS += CFLAGS="$(CFLAGS) -pthread"
COMPONENT_BUILD_ARGS += LDFLAGS="$(LDFLAGS)"
...
COMPONENT_INSTALL_ARGS += bindir=$(USRBINDIR.$(BITS))
COMPONENT_INSTALL_ARGS += includedir=$(USRINCDIR)
COMPONENT_INSTALL_ARGS += libdir=$(USRLIBDIR.$(BITS))
COMPONENT_INSTALL_ARGS += mandir=$(USRSHAREMANDIR)
```

chincheta0815 commented:
> 
> 
> @chincheta0815 I mean this: [#4726 (review)](https://github.com/OpenIndiana/oi-userland/pull/4726#pullrequestreview-195881928). It's suggested to be like this:
> 
> ```makefile
> COMPONENT_BUILD_ARGS += CC="$(CC)"
> COMPONENT_BUILD_ARGS += CFLAGS="$(CFLAGS) -pthread"
> COMPONENT_BUILD_ARGS += LDFLAGS="$(LDFLAGS)"
> ...
> COMPONENT_INSTALL_ARGS += bindir=$(USRBINDIR.$(BITS))
> COMPONENT_INSTALL_ARGS += includedir=$(USRINCDIR)
> COMPONENT_INSTALL_ARGS += libdir=$(USRLIBDIR.$(BITS))
> COMPONENT_INSTALL_ARGS += mandir=$(USRSHAREMANDIR)
> ```

okay now? all the requests changes etc. were a little bit confusing...
hope everything is okay now.
after all is beautiful, correct and checked I would then squash...

Mno-hime commented:
I think you can squash.

@chincheta0815 What could be done next time make this process less confusing?

pyhalov commented on components/database/lmdb/Makefile:
```diff
@@ -0,0 +1,55 @@
+COMPONENT_SUMMARY =				Lightning memory-mapped database library
+COMPONENT_SRC =					lmdb-LMDB_$(COMPONENT_VERSION)
+COMPONENT_ARCHIVE =				LMDB_$(COMPONENT_VERSION).tar.gz
+COMPONENT_ARCHIVE_HASH =		sha256:abf42e91f046787ed642d9eb21812a5c473f3ba5854124484d16eadbe0aa9c81
```
Can you make intendation look consistent?

pyhalov commented on components/database/lmdb/Makefile:
```diff
@@ -0,0 +1,55 @@
+include $(WS_MAKE_RULES)/justmake.mk
+include $(WS_MAKE_RULES)/ips.mk
+
+# set SOVERSION here since it is needed by all targets.
```
1) Can we set in in COMPONENT_BUILD_ENV and COMPONENT_INSTALL_ENV, as it's a common pattern instead of exporting?
2) What are lmdb promises about ABI stability? If you make SOVERSION=$(COMPONENT_VERSION), even minor update will result in ABI breakage.

chincheta0815 replied:
> 
> 
>     1. Can we set in in COMPONENT_BUILD_ENV and COMPONENT_INSTALL_ENV, as it's a common pattern instead of exporting?
Well, I thought I got the approval by Mno-hime. You are right. Shall I now again change something? --> changed it.
> 
>     2. What are lmdb promises about ABI stability? If you make SOVERSION=$(COMPONENT_VERSION), even minor update will result in ABI breakage.
The thing with the versioning was done due to the info I got from Mno-hime. He sent me info how OpenSuse and Fedora do the versioning. I just compied that. Concerning the ABI stability I cannot tell you. I am a simple user...

Mno-hime replied:
ABI stability looks good: https://abi-laboratory.pro/index.php?view=timeline&l=lmdb.

openSUSE uses version-based soname because they are aggressive in rebuilding, Fedora isn't that much so they stick with soname set to `0.0.0`. Maybe we should as well.

chincheta0815 replied:
Say I this way: tell me I change it to whatever you like. ;o) In the end I would say there is not real issue in using the complete version number and do a symlink as 0.0.0 always looks like a bug to me.

Mno-hime replied:
I suggest to follow https://src.fedoraproject.org/rpms/lmdb/blob/master/f/lmdb-make.patch in this.

chincheta0815 commented:
Was that squash okay?

chincheta0815 commented:
added the change of the SOVERSION handling.

chincheta0815 commented:
@Mno-hime I squashed after I added some comment by pyhalov. I hope everything is now okay. The thing making the process a little less confusing refers on the one hand to the docs. There are some in the wiki and some in the repo. They are hardly the same. Maybe a note which ones are more up-to-date would be good as some components are quite. Their first version differs due to the difference in the tools. Another thing is the input I got from all of you guys. Very helpful, indeed. Unfortunately, you seem to have different opinions. So I have to cope with both, making it a little confusing, e.g. the SOVERSION variable. Concerning things like ABI compatibility I am simply a "user" implementing the Makefiles etc. from the software devs, that is kind hard... In general I totally understand every question etc. due to the quality you want to achieve, but it's hard.

Mno-hime commented:
@chincheta0815 Sorry, for this to be confusing. I think you do great. In any case for your contributions!

About the process here in the PR: no one knows, what is the "right" answer, how to do this "right". However, we strive to get to the best possible outcome by discussion and hopefully drifting towards the best approach.

chincheta0815 commented:
@Mno-hime Thanks. I know that you do the best! And I really appreciate your work.

pyhalov commented:
Second publish fails:
```
$ gmake clean publish
$ touch Makefile
$ gmake publish
...
ln -s ./liblmdb-0.9.23.so /export/home/alp/srcs/tests/oi-userland/components/database/lmdb/build/prototype/i386/usr/lib/liblmdb.so
ln: cannot create /export/home/alp/srcs/tests/oi-userland/components/database/lmdb/build/prototype/i386/usr/lib/liblmdb.so: File exists
make[1]: *** [Makefile:46: install] Error 2
```
Likely, "ln -s" should be changed to "ln -sf".

chincheta0815 commented:
> Likely, "ln -s" should be changed to "ln -sf".

Should be okay now.

pyhalov commented:
sorry, one more issue - gmake test fails because CC,CFLAGS,LDFLAGS are not exported on gmake test.

chincheta0815 commented:
> 
> 
> sorry, one more issue - gmake test fails because CC,CFLAGS,LDFLAGS are not exported on gmake test.

No problem. Changed. I wonder why I did not get this issues. Maybe due to an overloaded env...
